### DISCUSSION TITLE
ColorLogic colors and shows not working properly

### DISCUSSION THREAD

**Discussion by Crewski** [tagyoureit/nodejs-poolController#476] [Nixie] [answered]
I have the Nixie controller setup to control my ColorLogic pool lights via a GPIO relay.  The relay is located on the high voltage side (before the transformer).  It is also missing "Voodoo Lounge" which would be sequence 1.  The relay is quickly toggling on/off, but the final color/show doesn't match up with what Nixie was trying to do.

This obviously is super low priority....it's not like I'm changing colors all the time (or even being in the pool at night for that matter).

From the documentation, in order to sync up the lights you need to turn in on, off for 11-15 seconds, then back on.  This resets to Voodoo Lounge (the first show in the sequence).  It also says that if it isn't cycling colors to leave it off for at least 1 second between cycles.  I'm not sure if this is needed or not when the relay is quickly switching on and off.

Let me know if there is anything you want me to try.  Thanks.

I'm on 26ff722 of njsPC and ae1b006 of REM.

**Comment by rstrouse** [comment]
Please pull njsPC, REM, and dashPanel.  I made some changes for the ColorLogic lighting themes, added Voodoo Lounge, and added a Sync button.  

Unfortunately, I do not have any of these lights so I will need you to do some field testing for me.  The docs for all these lights are often not up to par.  For instance, the ColorLogic docs do not say whether the sequence should start in the off position or the on position.  I suppose if you could witness the lights you would get a pretty good idea right away.

I don't know how noticeable the differences between the themes are but if we are one click off either way.  However, Cloud White is sandwiched in between Emerald and Warm Red so if you select Cloud white and the color comes out green or red we should be able to determine if the sequence is off and which direction it is off.

Finally, I may need you to fiddle on the timing.  Right now we delay the relay changes 100ms but that is based upon the acceptable timing for IntelliBrite.  I assume it cannot be faster so adding some delay in REM for the on and off sequences may determine if there is a need for additional delay.   I have to say I can't imagine that the switch controller they sell would require more than 17 seconds to switch a light theme.   While you are waiting for a change like that it is an eternity.  But pool lighting in general is pretty brain-dead.

**Reply by Crewski** [reply]
I got everything updated, I'll test things out this evening and report back.

How do I adjust the timing in REM?

Yeah, I doubt they would go that route as well as it would take forever.  The sad part is I actually had a controller, but it was in a box in my shed that a raccoon decided was going to be her "bathroom box" so it got thrown away.

**Reply by Crewski** [reply]
My assumption is that the controller keeps track of what the lights are set to.  So once they are synced, it'll stay that way unless there is a power outage maybe.  It probably compares what color its on to what color it wants to go to, then does the on/off sequence to get there.  If the new color is earlier in the list, it would cycles all the way through back to the beginning.  I'm just guessing on this of course.  I should be able to test it by picking lower sequence numbers and comparing them.

**Reply by rstrouse** [reply]
Nope your assumption is not correct as I understand it.  The sequence to get to Cabaret is to on-off 17 times.  The controller itself does not report anything so the only time the sequence can get out of sync is if you do not wait 10 seconds to change a theme between a theme request.   Or you change the theme on the light from some other source.

If there is a power outage that happens for 11-15 seconds then comes back on, it will reset to Voodoo Lounge though.

It doesn't look like we have the ability to tweak the relay timing on GPIO relays.  This function only exists on I2c relays.

**Reply by Crewski** [reply]
I did some quick testing tonight.
1 - If the lights are on and you hit the sync button, it properly resets back to voodoo lounge.
2 - If the lights are on, on voodoo lounge, and you press Cloud White, it ends up on Aqua Green (2 colors off).
3 - The sequencing seems to be too fast.  The higher the sequence number, the farther off the final color/show seems to be.
4 - When the light has been off for 60 seconds, it'll start up white for 15 seconds in order to see the pool well.  If any color/show/sync is selected when the light has been off for 60 seconds, it'll just keep flashing white and then stay at whatever color/show it started at.  Colors can't be changed during this 15 second "startup".
5 - Pressing Deep Blue does not go to that color, it just advances to the next color since the sequence number is two.  Based on this, I would think either a very long reset must happen first (to account for the white light potential startup and the reset) then go to the color, or compare the selected show to the current show and make the change from that starting point (again taking into account for the white startup).

I tried to find some "user video" of people with the ColorLogic Controller to hopefully see what is going on with the lights while it is in use, but I couldn't find any.  Maybe I can see if someone on trouble free pool has one that could take a video or something to show us what is happening.

**Reply by Crewski** [reply]
I modified the on/off sequence function and will try it out tonight.  Hopefully it'll give me a starting point of something that works and then tune the timing from there.  The key changes are comparing sequences and only moving the difference between them, ability to change the timeout for isOn: true and isOn: false, and a safety light delay if the light is off.

```
public async sendOnOffSequenceAsync(count: number | { isOn: boolean, timeout: number }[], timeout?: number): Promise<InterfaceServerResponse> {
        try {
            this._sequencing = true;
            let arr = [];
            let cstate = state.circuits.getItemById(this.circuit.id);
            if (typeof count === 'number') {
                let tOn = typeof timeout === 'undefined' ? 100 : timeout;
                let tOff = typeof timeout === 'undefined' ? 100 : timeout;                
                if (cstate.type == 20){   // check to see if light is colorlogic (20) 
                    tOn = typeof timeout === 'undefined' ? 1000 : timeout;  // change delay after on
                    tOff = typeof timeout === 'undefined' ? 1000 : timeout; // change delay after off      
                    let thm = sys.board.valueMaps.lightThemes.findItem(cstate.lightingTheme);  // get the previous theme
                    count = count - thm.sequence + 1;   // compare sequences and advance                    
                    if (count < 0) count = count + 17;  // if the new theme is earlier in the sequence we must loop all the way through
                }
                else if (cstate.isOn) arr.push({ isOn: false, timeout: 1000 });
                //arr.push({ isOn: false, timeout: t }); // This may not be needed but we always need to start from off.
                //[{ isOn: true, timeout: 1000 }, { isOn: false, timeout: 1000 }]
                for (let i = 0; i < count; i++) {
                    if (i < count - 1) {
                        arr.push({ isOn: true, timeout: tOn });
                        arr.push({ isOn: false, timeout: tOff });
                    }
                    else arr.push({ isOn: true, timeout: 1000 });
                }
                console.log(arr);
            }
            else arr = count;
            if (cstate.type == 20 && !cstate.isOn) arr.unshift({ isOn: true, timeout: 16000});    // colorlogic: we need to wait for the safety light to clear
```

**Comment by rstrouse** [comment]
Go ahead and pull REM.  Lets get the sequences to work properly first.  I added the ability to insert additional delay time on the GPIO pin for sequences.  When we get this figured out I will make it so the 15 second on delay takes effect.  What an odd feature.  It's bad enough that the light engine isn't responsive enough for a 300ms delay between clicks.  There is a potential for 17 stinking clicks.  
![image](https://user-images.githubusercontent.com/47839015/172737566-dd22f78a-ef55-45f6-aa9a-79743be88d95.png)



If you are standing outside at night in a wet swimsuit imagine turning on the light then waiting 15 seconds for the white light to stop.  Then after that you could potentially have to flip that stupid switch like a robot 17 times.  Wtf are they thinking!?  All told you are a couple of minutes into it just to get the lights on.  At that point I would be oh hell no let's swim in the dark or just take whatever it gives me.

**Comment by rstrouse** [comment]
Hey have a look at this manual excerpt it is showing only 12 themes and I keep seeing two versions of the lights being referenced in various places.  There appears to be a v2.5 and a v4.0 that have differing settings.  Even the most basic of choices are different than the 17.
![image](https://user-images.githubusercontent.com/47839015/172748674-f70d25e7-fcba-487f-8e4c-b10a0d75cf23.png)

The other thing you should try is when you have reset the lights so that it is running voodoo, do you get Deep Sea Blue if you click the button in REM off for 1 second then back on?

**Comment by Crewski** [comment]
Thanks for adding the delays in REM, that helped out a lot.  I just got done running through bunch of testing.  Let's start with delays.  In REM I added 100ms to both the on and off.  With that set, I was able to reliably get the lights to respond as intended.  I tried dropping down to 50 but that was causing issues.  The extra 100ms is probably on the edge but seemed to work just fine.

If on Voodoo Lounge, toggling the relay off for 1 second in REM does advance to Deep Blue.

Using the code as is and starting at Voodoo Lounge with the light on, selecting a color/theme caused the lights to overshoot by 1.  So Cloud White actually ended on Warm Red.  This is caused by the initial OFF command you have in the for intellibrite (based on comments)  After that first selection, anything after that was wrong.  These lights just continue to cycle through from wherever they are, they don't reset back to the beginning unless you run the sync procedure.

I modified the code to check where the lights currently were then figure out how far it had to go.  This code along with the extra 100ms worked reliably when starting with the light on (after the safety white light time had passed).  I was able to pick any color/show and it would end on it.  It would cycle completely through if it had to.  Speak of the safety light, the documentation is wrong.  It's not 60 seconds of it being off, it's anything passed the reset time.  So anytime you turn off the light, when you turn it back on your only options are advance, reset, or safety light for 15s then to the color it was on.  So dumb.  Anyway, below is the code that worked great.  Only thing left is handling that dumb safety light.

```
    public async sendOnOffSequenceAsync(count: number | { isOn: boolean, timeout: number }[], timeout?: number): Promise<InterfaceServerResponse> {
        try {
            this._sequencing = true;
            let arr = [];
            let cstate = state.circuits.getItemById(this.circuit.id);
            if (typeof count === 'number') {
                let t = typeof timeout === 'undefined' ? 100 : timeout;              
                if (cstate.data.type.val  == 14){   // check to see if light is colorlogic  
                    let thm = sys.board.valueMaps.lightThemes.findItem(cstate.lightingTheme);  // get the previous theme
                    count = count - thm.sequence + 1;   // compare sequences and advance                
                    if (count < 1) count = count + 17;  // if the new theme is earlier in the sequence we must loop all the way through     
                }
                else if (cstate.isOn) arr.push({ isOn: false, timeout: 1000 });
                //arr.push({ isOn: false, timeout: t }); // This may not be needed but we always need to start from off.
                //[{ isOn: true, timeout: 1000 }, { isOn: false, timeout: 1000 }]
                for (let i = 0; i < count; i++) {
                    if (i < count - 1) {
                        arr.push({ isOn: true, timeout: t });
                        arr.push({ isOn: false, timeout: t });
                    }
                    else arr.push({ isOn: true, timeout: 1000 });
                }
                console.log(arr);
            }
            else arr = count;
            // The documentation for IntelliBrite is incorrect.  The sequence below will give us Party mode.
            // Party mode:2
            // Start: Off
            // On
            // Off
            // On
            // According to the docs this is the sequence they lay out.
            // Party mode:2
            // Start: On
            // Off
            // On
            // Off
            // On

            let res = await NixieEquipment.putDeviceService(this.circuit.connectionId, `/state/device/${this.circuit.deviceBinding}`, arr, 60000);
            // Even though we ended with on we need to make sure that the relay stays on now that we are done.
            if (!res.error) {
                this._sequencing = false;
                await this.setCircuitStateAsync(cstate, true, false);
            }
            return res;
        } catch (err) { logger.error(`Nixie: Error sending circuit sequence ${this.id}: ${count}`); }
        finally { this._sequencing = false; }
    }
```

My light are the "newer" version and have the 17 shows, same as you have programmed.
![image](https://user-images.githubusercontent.com/8012153/172751378-08c8b3e1-9fdb-400a-835a-381e8dd6c362.png)

**Reply by rstrouse** [reply]
By chance do you know if they loop around for instance if it is on Cabaret and you advance to the next one will it go to Voodoo Lounge?

**Reply by Crewski** [reply]
Yeah, that's how it work.

**Comment by rstrouse** [comment]
The safety light thingy is equivalent to a light command we simply need to set the state so that the options to perform anything during this period is suspended.  So are you saying that if the light has been off for 15 seconds it will go into safety light mode?  I'd love to party with the design engineer.  I picture Dr. Emmett Brown with a whole lotta LSD on board.  It would be fun to watch.

**Reply by Crewski** [reply]
Yes, if the light has been off for 15 seconds (I didn't count the exact seconds, but it was definitely more than 15 and less than 30) it goes into the safety light mode.  So for the off timing we get:
0-10: advance to next show
11-15: reset to Voodoo Lounge
15+: shine 1.21 gigawatts of white beauty into the pool.

EDIT: I went out and counted, it definitely is anything after 15 seconds.

**Comment by rstrouse** [comment]
Pull njsPC and dashPanel.  I had to make a specific theme setting algorithm that marshals the theme process to the specific light brand.  So the code for setting the colorLogic themes are in the setColorLogicThemeAsync method.  This should accommodate your sequence driver as well as deal with users turning the lights on and off indiscriminately.  If the user turns on and off the light within 10 seconds then it will advance to the next theme.  If the circuit has been off for more than 15 seconds then it will first display as white and then after 15 seconds show the current theme.   During our white phase the theme setting abilities are disabled.  You cannot change the theme.

Too bad Daniel Coon Caddyshacked your switch controller.  It would have been interesting to see what the LEDs did during these sequences.

**Comment by Crewski** [comment]
We are so close!  The cycling through seems to be working great.  The only issue that remains is due to bad documentation of the safety light.  Its supposed to be on for 15 seconds, but it is more like 16 seconds.  So as the code is now, it starts cycling a little bit too early so it gets out of sync.  Playing with the delays, I was able to get it to work with an additional 750ms....but that was probably living dangerously because I couldn't really see the safety light go away before the cycling happened.  I'd suggest rounding up and making it a 16000ms delay.  What's another 250ms when you are already patiently waiting for 15750ms to pass by.

**Reply by rstrouse** [reply]
Did you make the changes in the code and try it out?  I assume Circuit.ts:177 is what you changed.  I changed that line to hold on for 16000ms.

**Reply by Crewski** [reply]
No, I changed the delay on in REM.  When using a stop watch, I was getting around 16.5 seconds before the safety light turns off, so that makes more sense if I was starting with 16000ms with an additional 750-1000ms delay.

**Comment by rstrouse** [comment]
Go ahead and pull it and see if we have enough safety dance.

**Reply by Crewski** [reply]
I tried it last night but there was some issues.  I was having some weird stuff going on anyway, so I'll reboot everything and try again this weekend.  I'll let you know.

**Reply by rstrouse** [reply]
@Crewski how did your testing work out?

**Reply by Crewski** [reply]
@rstrouse Thanks for following up.  I got derailed a little with a bad raspberry pi and then some mqtt issues.

I just tested it out.  Starting with the light off and selecting Cloud White, it ended up on Aqua Green (which is 2 away from the target).  I went into the code and changed it to a 17 second delay (what's another second at this point) and it seems to work well.  I tested several times and it landed correctly on Cloud White each time.

However, there is a new bug that has been introduced.  Starting with the light off and selecting a color theme with the extra second works as expected.  However a few seconds later njsPC seems to go through a sequence and resets back to Voodoo Lounge.  This is only in the code, the actual pool light does not change and remains on Cloud White.  It happens every time.  If I turn the light off after landing on cloud white, the button on dashPanel is flashing yellow as if it was being pushed to go back through the sequence to get to Voodoo Lounge.  If I leave the light on, it'll just change itself back to Voodoo Lounge.  I've attached a replay of the stated scenario.  Thanks.
[replay (16).zip](https://github.com/tagyoureit/nodejs-poolController/files/8985842/replay.16.zip)

**Reply by rstrouse** [reply]
Just to clarify: If you turn the light back on after it has been off for greater than 10 seconds and less than 15 seconds, doesn't the light reset to Voodoo lounge?

**Comment by rstrouse** [comment, marked as answer]
@Crewski pull njsPC I figured out that it wasn't necessarily Voodoo lounge that it was being reset to, the issue was that the final on was kicking off the safety dance some crazy thing about some dudes that don't wear hats.  Sorry about the outer joins but you need to get in a mood to deal with pool lighting.

**Reply by Crewski** [reply]
Alright, looks like you got everything working properly.

- The 17 second delay for the "safety light" works well.
- njsPC doesn't reset the pool light theme like before.
- Picking themes whether the light is on or off works great.
- Manually turning the light on and off advances the color.
- Manually turning the light on after 12-ish second does properly reset things to Voodoo Lounge

As far as I can tell you can consider ColorLogic lighting complete!  Thank you so much for sorting this all out.
